Cache iterator next method fix:#5860

rhuanjl opened the pull request:
Per spec when using GetIterator the .next() method should be cached.
See step 6 at: https://tc39.es/ecma262/#sec-getiterator  

Currently this is not done - instead CC uses Get to access .next() for every time it's called - these Get calls are observable and incorrect per spec - they're also inefficient. This PR aims to bring CC inline with spec on this point.

There are broadly 3 ways in which GetIterator and .next() are used which in turn require 3 different fixes:

1. In library methods implemented in C++ These are fixed by cacheing the .next() method before its first use in a local variable.

2. In facilities implemented in bytecode e.g. for...of, and the spread operator. This was fixed in #6312 this PR adds tests for these usages.

3. In yield* This was fixed in #6312 this PR adds tests for this usage.

**Extra note:** this will fix the following test262 test which currently fails: https://github.com/tc39/test262/blob/master/test/language/statements/for-of/iterator-next-reference.js

Fix:#5860

rhuanjl commented on lib/Runtime/Language/InterpreterStackFrame.cpp:
```diff
@@ -430,15 +430,17 @@
 
-#define PROCESS_A2NonVarToA1Reg_COMMON(name, func, suffix) \
+#define PROCESS_A3NonVarToA1Reg_COMMON(name, func, suffix) \
     case OpCode::name: \
```
New layout to support ResumeYieldStar with a cached nextMethod.

rhuanjl commented:
This will clearly need some sort of update to be compatible with #6170 I'll re-visit it when that is merged.

rhuanjl commented:
I've updated this so it no longer conflicts with #6170. Please could someone take a look at this now?

zenparsing commented on lib/Runtime/Language/JavascriptOperators.cpp:
```diff
@@ -11251,27 +11262,32 @@ using namespace Js;
+    // IteratorNext as described in ES6.0 (draft 22) Section 7.4.2
+    RecyclableObject* JavascriptOperators::IteratorNext(RecyclableObject* iterator, ScriptContext* scriptContext, RecyclableObject* nextFunc, Var value)
+    {
+        ThreadContext *threadContext = scriptContext->GetThreadContext();
```
Nit:

```suggestion
        ThreadContext* threadContext = scriptContext->GetThreadContext();
```

zenparsing replied:
Meant to add this as part of a larger review - more incoming! : )

zenparsing commented on lib/Runtime/ByteCode/ByteCodeEmitter.cpp:
```diff
@@ -10302,7 +10334,15 @@ void EmitYieldStar(ParseNodeUni* yieldStarNode, ByteCodeGenerator* byteCodeGener
+    uint cacheId = funcInfo->FindOrAddInlineCacheId(iteratorObject, Js::PropertyIds::_symbolIterator, false, true);
+    byteCodeGenerator->Writer()->PatchableProperty(Js::OpCode::StFld, iteratorLocation, iteratorObject, cacheId);
+    cacheId = funcInfo->FindOrAddInlineCacheId(iteratorObject, Js::PropertyIds::next, false, true);
+    byteCodeGenerator->Writer()->PatchableProperty(Js::OpCode::StFld, nextCache, iteratorObject, cacheId);
```
If I understand correctly, this will create a temporary object whose `Symbol.iterator` property will be the "real" iterator object and whose `next` property is the cached next method. And this object is ultimately used in `JavascriptOperators::OP_ResumeYield`.

A couple of thoughts:

- Symbol.iterator would need to be a method (if it's used at all by `OP_ResumeYield`).
- Will calling the next method from `OP_ResumeYield` result in the correct this value? Or will the this value be the wrapper iterator?
- Does `OP_ResumeYield` also expect "throw" and "return" to be present on the wrapper iterator (if they exist on the original iterator)?

Let's think about how we would want this to work ideally. Perhaps `OP_ResumeYield` should expect an object of the following form instead?

```
{
  iterator: Var; // The yield star iterator
  next: JavascriptFunction* // The cached next method
}
```

@boingoing Thoughts?

rhuanjl replied:
It uses symbol.iterator as a convenient label the object it puts there is the yield* iterator and not the orriginal Symbol.iterator

Over in ResumeYield it is retrieved as iteratorObject and correctly used as the this value, it is not called.

rhuanjl commented:
I've updated to latest and renamed the iterator object to get rid of the confusion.

I still really don't like how ResumeYield works but seems like about the best fix to this error I  can do without adding extra bytecode instructions.

I'm somewhat confused by the test failure it doesn't seem related to this.

rhuanjl commented:
@zenparsing I've updated to latest and regenerated the bytecode again and it now passes all tests; do you think this could be merged?

rhuanjl commented:
I've updated the initial comment to explain how this fix works more clearly. I'll hold on updating the bytecode again until you've had time to review again @zenparsing

rhuanjl commented:
Is there anything I can do to help get this landed?

rhuanjl commented:
I've updated this to work with #6312 (which I've now merged) it fixed this issue for generator functions and for iterator calls emitted in bytecode, so I've removed those components from this PR.

This PR is now limited to updating iterator usage within library functions.

@ppenzin @chicoxyzzy this fixes a long standing spec deviation in CC - I wrote this about a year ago though it was stalled due to clashing with #6312

rhuanjl commented:
@Fly-Style @penzn please could one of you review this - it's nowhere near of complex as it looks - most of the diff is new tests.

To summarise:
1. Per spec iterator next() methods should be cached so if there's an accessor for the next() method it should only be called once
2. CC used to call the accessor every time next() was called
3. #6312 which has been merged fixed this in some cases (generators and any iterators embedded in bytecode)
4. This PR fixes this for iterators used within library methods AND adds tests

ppenzin commented:
It took me a little bit to get through this, but wasn't too bad.

rhuanjl commented:
So good to finally get this in, thanks @ppenzin

ppenzin commented:
@rhuanjl, you are welcome, looking forward to closing off some other long-running issues.
